### PR TITLE
Bug #10794:

### PR DESCRIPTION
**Pull request by rohitkeshri1986**
Added UI to System -> Boot which allows the user to tag a Boot Environment as "Keep"
As keep tag is set it calls
	be = Update.FindClone("default");
	Update.CloneSetAttr(be, keep = True) and handled accordingly.

### CODE REVIEW

File: gui/templates/system/bootenv_keep.html [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{% extends "freeadmin/generic_form.html" %}
+
+{% block form %}
+<tr><td colspan="2"><span>Do You Want To Keep BootEnv</span></td></tr>
```
Comment (williambr) [inline gui/templates/system/bootenv_keep.html]: Please use translation here, also do not use upper case for every word.

{% trans "Do you want...." %}

**Comment by williambr** [comment]
You have introduced 3 new flake8 errors: https://travis-ci.org/freenas/freenas/builds/127714270

Also, I believe there should be a way to see the boot environments with Keep flag set from the datagrid.

**Comment by jordanhubbard** [comment]
Rohit - any comments on the Travis CI errors that were flagged?  Looks like this is really close, but still not passing the flake checks and there's another UI feature still missing, namely, setting the keep flag from the data grid.

**Comment by rohitkeshri1986** [comment]
1. Jordan - As asked for " setting the keep flag from the data grid." 
To address this, i am planning to do following changes as shown in attached screenshot. Please review 
and give your comments. 
2. Regarding  3 new flake8 errors: -  Could you please explain what it is as this was generated at the time when i created the pull request.
3. William - I have fixed the changes for html and working on GUI enhancement for to see the boot environments with Keep flag set from the datagrid. Please review the attachment for GUI that i am working upon.
keep-boot-2

**Comment by jordanhubbard** [comment]
1. Looks good to me!  I like it.
2. This was auto-reported by Travis CI (https://travis-ci.org/freenas/freenas/builds/127714270); see William's comment above.  Should be easy to fix, if you have not already (if you have, just submit another pull request when you have this done and Travis CI will either indicate no errors or check again and flag the errors).  You can also run flake yourself to validate that the code is PEP8 compliant.

**Comment by rohitkeshri1986** [comment]
We should have a CloneGetValAttr(clone) that would give us the VALUE for a PROPERTY.
----------------------------------------------------------------------
# zfs get beadm:keep freenas-boot/ROOT/Initial-Install
NAME                               PROPERTY    VALUE       SOURCE
freenas-boot/ROOT/Initial-Install  beadm:keep  True        local
----------------------------------------------------------------------
# zfs get -H -o value beadm:keep freenas-boot/ROOT/Initial-Install
True
----------------------------------------------------------------------
With this API we can get the status of a BE and use it while setting the keep flag into the data grid.
This API can be like as below.
def CloneGetValAttr(clone):
    """
    Given a clone, get value as an attributes.
    Returns True if the BootEnv is Kept
    False otherwise if not Kept.
    """

    # This maps to zfs get -H -o value beadm:keep freenas-boot/ROOT/${bename}
    cmd = ["/sbin/zfs", "get", "-H", "-o", "value", "beadm:keep", "freenas-boot/ROOT/%s" % clone["realname"]]
    ret = subprocess.check_call(cmd)
    return ret

Please suggest for the same.

**Comment by williambr** [comment]
Sure, that works. Just add in the PR.

On Tue, May 10, 2016 at 9:49 AM, rohitkeshri1986 <notifications@github.com>
wrote:

> We should have a CloneGetValAttr(clone) that would give us the VALUE for a
> PROPERTY. zfs get beadm:keep freenas-boot/ROOT/Initial-Install
>
> NAME PROPERTY VALUE SOURCE
> freenas-boot/ROOT/Initial-Install beadm:keep True local zfs get -H -o
> value beadm:keep freenas-boot/ROOT/Initial-Install True
>
> With this API we can get the status of a BE and use it while setting the
> keep flag into the data grid.
> This API can be like as below.
> def CloneGetValAttr(clone):
> """
> Given a clone, get value as an attributes.
> Returns True if the BootEnv is Kept
> False otherwise if not Kept.
> """
>
> # This maps to zfs get -H -o value beadm:keep freenas-boot/ROOT/${bename}
> cmd = ["/sbin/zfs", "get", "-H", "-o", "value", "beadm:keep", "freenas-boot/ROOT/%s" % clone["realname"]]
> ret = subprocess.check_call(cmd)
> return ret
>
> Please suggest for the same.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/freenas/freenas/pull/144#issuecomment-218147148>
>



-- 
William Grzybowski / wg
FreeBSD Ports Committer

**Comment by rohitkeshri1986** [comment]
We should have a CloneGetValAttr(clone) that would give us the VALUE for a PROPERTY.
----------------------------------------------------------------------
# zfs get beadm:keep freenas-boot/ROOT/Initial-Install
NAME                               PROPERTY    VALUE       SOURCE
freenas-boot/ROOT/Initial-Install  beadm:keep  True        local
----------------------------------------------------------------------
# zfs get -H -o value beadm:keep freenas-boot/ROOT/Initial-Install
True
----------------------------------------------------------------------
With this API we can get the status of a BE and use it while setting the keep flag into the data grid.
This API can be like as below.
def CloneGetValAttr(clone):
    """
    Given a clone, get value as an attributes.
    Returns True if the BootEnv is Kept
    False otherwise if not Kept.
    """

    # This maps to zfs get -H -o value beadm:keep freenas-boot/ROOT/${bename}
    cmd = ["/sbin/zfs", "get", "-H", "-o", "value", "beadm:keep", "freenas-boot/ROOT/%s" % clone["realname"]]
    ret = subprocess.check_call(cmd)
    return ret

Please suggest for the same

**Comment by williambr** [comment]
Sure, that works. Just add in the PR.

**Comment by williambr** [comment]
You can also make that method name more generic and accept the attribute name.

**Comment by rohitkeshri1986** [comment]
I found an issue with ListClones module code. ListClones fetches keep information from ZFS.
Irrespective of any value for "keep" it always shows as "True" in dictionary that it returns.

Existing Code:
--------------
try:
prop_cmd = ["/sbin/zfs", "get", "-H", "-o", "value", "beadm:keep"]
prop_cmd.append("freenas-boot/ROOT/{0}".format(tdict["realname"]))
keep_str = subprocess.check_output(prop_cmd).rstrip()
if keep_str == "-":
tdict["keep"] = None
else:
tdict["keep"] = bool(keep_str)
except:
log.debug("Could not get beadm properties from dataset")
rv.append(tdict)
return rv

Proposed Code:
--------------
try:
prop_cmd = ["/sbin/zfs", "get", "-H", "-o", "value", "beadm:keep"]
prop_cmd.append("freenas-boot/ROOT/{0}".format(tdict["realname"]))
keep_str = subprocess.check_output(prop_cmd).rstrip()
if keep_str == "-":
tdict["keep"] = None
else:
if keep_str == "True":
tdict["keep"] = True
elif keep_str == "False":
tdict["keep"] = False
else:
tdict["keep"] = None
except:
log.debug("Could not get beadm properties from dataset")
rv.append(tdict)
return rv

**Comment by williambr** [comment]
Yes, that looks fine.

**Comment by williambr** [comment]
You have introduced new flake8 errors, check https://travis-ci.org/freenas/freenas/builds/130818511

**Comment by rohitkeshri1986** [comment]
How can i fix it.
It came in my last push too.
Please suggest.

**Comment by williambr** [comment]
The new introduced errors are identified in the link I pasted you:

+gui/system/views.py:440:37: E251 unexpected spaces around keyword / parameter equals
+gui/system/views.py:440:39: E251 unexpected spaces around keyword / parameter equals
+gui/system/views.py:456:1: E303 too many blank lines (3)

File: gui/templates/system/bootenv_keep.html [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{% extends "freeadmin/generic_form.html" %}
+
+{% block form %}
+<tr><td colspan="2"><p translate="no">Do you want to keep bootenv?</p></td></tr>
```
Comment (williambr) [inline gui/templates/system/bootenv_keep.html]: <p translate="no"> ? Makes no sense here

Please use:

{% trans "Do you want to set keep flag in this boot environment?" %}

**Comment by williambr** [comment]
You have introduced a new error :)

+gui/system/views.py:439:29: W291 trailing whitespace

Please try to use a pep8/flake8 analyzer in your text editor.

**Comment by williambr** [comment]
Also, now that I think about it. How would one unset a keep flag for BE?

**Comment by rohitkeshri1986** [comment]
should we add a unkeep kind of interface too.

**Comment by williambr** [comment]
Sure, adding keep without the ability to undo it doesnt make much sense.
Thanks!

**Comment by rohitkeshri1986** [comment]
Added unkeep support to GUI.
Fixed flake8 errors.

**Comment by rohitkeshri1986** [comment]
gui/system/alertmods/selftest.py:8:1: E302 expected 2 blank lines, found 1
 gui/system/alertmods/vmwaresnap_fails.py:16:44: F841 local variable 'lock' is assigned to but never used
 gui/system/alertmods/vmwaresnapdelete_fails.py:16:54: F841 local variable 'lock' is assigned to but never used
+gui/system/views.py:455:1: E303 too many blank lines (3)
+gui/system/views.py:1755:23: F821 undefined name 'combined'
+gui/system/views.py:1757:38: F821 undefined name 'combined'
 gui/system/ixselftests/TestStatus.py:3:1: F401 'sys' imported but unused
 gui/system/ixselftests/TestStatus.py:3:10: E401 multiple imports on one line
 gui/system/ixselftests/TestStatus.py:5:1: F401 'TEST_CRITICAL' imported but unused

File: gui/templates/system/bootenv_unkeep.html [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{% extends "freeadmin/generic_form.html" %}
+
+{% block form %}
+{% trans "Do you want to reset keep flag in this boot environment?" %}
```
Comment (williambr) [inline gui/templates/system/bootenv_unkeep.html]: Why did you remove <tr><td colspan="2">?

File: gui/templates/system/bootenv_keep.html [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {% extends "freeadmin/generic_form.html" %}
 
 {% block form %}
-<tr><td colspan="2"><p translate="no">Do you want to keep bootenv?</p></td></tr>
+{% trans "Do you want to set keep flag in this boot environment?" %}
```
Comment (williambr) [inline gui/templates/system/bootenv_keep.html]: Why did you remove <tr><td colspan="2">?

**Comment by williambr** [comment]
I still see issues.

Keep button will show up even if selected BE already have keep flag set.
Unkeep button will show up even if selected BE does not have keep flag set.

**Comment by rohitkeshri1986** [comment]
Thank you for the comments.
I will look into it.

**Comment by rohitkeshri1986** [comment]
I have added keep/unkeep button hiding support.
Please review and suggest.
Thank you.

**Comment by williambr** [comment]
It looks good to me, but it dont actually works. I have this error when trying to access the datagrid:

```
May 27 04:44:46 freenas manage.py: [api.utils:12] Traceback (most recent call last):
May 27 04:44:46 freenas manage.py: [api.utils:12]   File "/usr/local/www/freenasUI/../freenasUI/api/utils.py", line 234, in dispatch
May 27 04:44:46 freenas manage.py: [api.utils:12]     request_type, request, *args, **kwargs
May 27 04:44:46 freenas manage.py: [api.utils:12]   File "/usr/local/lib/python2.7/site-packages/tastypie/resources.py", line 464, in dispatch
May 27 04:44:46 freenas manage.py: [api.utils:12]     response = method(request, **kwargs)
May 27 04:44:46 freenas manage.py: [api.utils:12]   File "/usr/local/www/freenasUI/../freenasUI/api/resources.py", line 2990, in get_list
May 27 04:44:46 freenas manage.py: [api.utils:12]     bundles.append(self.full_dehydrate(bundle))
May 27 04:44:46 freenas manage.py: [api.utils:12]   File "/usr/local/lib/python2.7/site-packages/tastypie/resources.py", line 852, in full_dehydrate
May 27 04:44:46 freenas manage.py: [api.utils:12]     bundle.data[field_name] = field_object.dehydrate(bundle, for_list=for_list)
May 27 04:44:46 freenas manage.py: [api.utils:12]   File "/usr/local/lib/python2.7/site-packages/tastypie/fields.py", line 130, in dehydrate
May 27 04:44:46 freenas manage.py: [api.utils:12]     raise ApiFieldError("The object '%r' has an empty attribute '%s' and doesn't allow a default or null value." % (previous_object, attr))
May 27 04:44:46 freenas manage.py: [api.utils:12] ApiFieldError: The object '<freenasUI.system.utils.BootEnv object at 0x814b76d90>' has an empty attribute 'keep' and doesn't allow a default or null value.
```

**Comment by rohitkeshri1986** [comment]
Thank you william.
I have tried my latest modificed code on multiple VM's, hope it should work fine.
Thank You.

**Comment by williambr** [comment]
That issue has been fixed but the feature is still not right.

I click Keep, it says it was successful but the datagrid says Keep is still No.

**Comment by rohitkeshri1986** [comment]
Please tell me the iso name on which you have tested.
I will test and fix. As i have tested it on FreeNAS-9.3-STABLE-201601181840.iso

**Comment by williambr** [comment]
I have tested it in latest 9.10 Nightlies. This is were all development should be done.

**Comment by rohitkeshri1986** [comment]
This problem is due to 
https://bugs.freenas.org/issues/15453

I have fixed and added it in this patch.

**Comment by williambr** [comment]
No, that change is already in master, you commit has just created a conflict.

**Comment by williambr** [comment]
The problem is that it is not setting the beadm:keep zfs property.

**Comment by rohitkeshri1986** [comment]
Hi William,
I have investigated the problem and have found out that CloneSetAttr for 9.10 has been modified and error is because of that.
when beadm:keep status is seen using command line for the first time its installed, it shows as below.
[root@freenas] /usr/local/www/freenasUI# zfs get -H -o value beadm:keep freenas-boot/ROOT/Initial-Install
-

Existing CloneSetAttr fails to set beadm:keep from value = "-" to "True/False" and so grid does not changes.
So, if we modify CloneSetAttr to below

    if "keep" in kwargs:
        tval = bool(kwargs["keep"])
        cmd = ["/sbin/zfs", "set", "beadm:keep=%s" % tval, "freenas-boot/ROOT/%s" % clone["realname"]]
        subprocess.check_call(cmd)
        return True
    return False

It is working fine.
This problem is seen only when beadm:keep value is -

**Comment by williambr** [comment]
Ok, I have committed a different fix to CloneSetAttr.

**Comment by williambr** [comment]
Merged, thanks.

**Comment by rohitkeshri1986** [comment]
Thank You.
